source/extensions/tracers/opentelemetry/http_trace_exporter.cc

```diff
@@ -0,0 +1,79 @@
+#include "http_trace_exporter.h"
+
+#include <chrono>
+#include <memory>
+#include <string>
+
+#include "source/common/common/logger.h"
+#include "source/common/protobuf/protobuf.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Tracers {
+namespace OpenTelemetry {
+
+OpenTelemetryHttpTraceExporter::OpenTelemetryHttpTraceExporter(
+    Upstream::ClusterManager& cluster_manager, envoy::config::core::v3::HttpService http_service,
+    OpenTelemetryTracerStats& tracing_stats)
+    : cluster_manager_(cluster_manager), http_service_(http_service),
+      tracing_stats_(tracing_stats) {}
+
+bool OpenTelemetryHttpTraceExporter::log(const ExportTraceServiceRequest& request) {
+
+  std::string request_body;
+
+  const auto ok = request.SerializeToString(&request_body);
+  if (!ok) {
+    ENVOY_LOG(warn, "Error while serializing the binary proto ExportTraceServiceRequest.");
+    return false;
+  }
+
+  const auto thread_local_cluster =
+      cluster_manager_.getThreadLocalCluster(http_service_.http_uri().cluster());
+  if (thread_local_cluster == nullptr) {
+    ENVOY_LOG(error, "OTLP HTTP exporter failed: [cluster = {}] is not configured",
+              http_service_.http_uri().cluster());
+    return false;
+  }
+
+  Http::RequestMessagePtr message = Http::Utility::prepareHeaders(http_service_.http_uri());
+
+  message->headers().setReferenceMethod(Http::Headers::get().MethodValues.Post);
+  message->headers().setReferenceContentType(Http::Headers::get().ContentTypeValues.Protobuf);
+
+  // add all custom headers to the request
+  for (const auto& header_value_option : http_service_.request_headers_to_add()) {
+    message->headers().setCopy(Http::LowerCaseString(header_value_option.header().key()),
+                               header_value_option.header().value());
+  }
+  message->body().add(request_body);
+
+  auto options = Http::AsyncClient::RequestOptions().setTimeout(std::chrono::milliseconds(
+      DurationUtil::durationToMilliseconds(http_service_.http_uri().timeout())));
+
+  Http::AsyncClient::Request* http_request =
+      thread_local_cluster->httpAsyncClient().send(std::move(message), *this, options);
+  tracing_stats_.http_reports_sent_.inc();
+
+  return http_request;
+}
+
+void OpenTelemetryHttpTraceExporter::onSuccess(const Http::AsyncClient::Request&,
+                                               Http::ResponseMessagePtr&& message) {
+  tracing_stats_.http_reports_success_.inc();
+  const auto response_code = message->headers().Status()->value().getStringView();
+  if (response_code != "200") {
+    ENVOY_LOG(warn, "response code: {}", response_code);
```

Comment:
Please elaborate why this is a warning (what is the expected behavior). Generally speaking a warning should give the operator enough context so they will not need to look at the source code to understand what is the problem.

Reply:
Hum I took this over from the original PR and haven't paid much attention. I think this should be an error log.

Anything over 200 is a error, and it can be many things. I think here the only thing we can do is say that there was a problem exporting the OTLP message. This should offer enough guidance for users to look at their OTel collectors logs or their telemetry back-end of choice. 


Expanding, the OTLP spec offers ways to deal with the failures https://github.com/open-telemetry/opentelemetry-proto/blob/main/docs/specification.md#failures, but if it's OK for you, I would address it as a follow up. The gRPC exporter also does not treat such errors following the spec today.